File: leetcode/balanced_number.cpp
class Solution {
    public:
        int nextBeautifulNumber(int n) {
            int* digits = new int[7]();
            int* counts = new int[10]();
            int i = 0;
            int d;
            int N = n;
            while (n > 0) {
                d = n % 10;
                digits[i] = d;
                counts[d]++;
                n = n / 10;
                i++;
            }
            increment(digits, counts);
            N++;
            while (!isValid(counts)) {
                increment(digits, counts);
                N++;
            }
            return N;
        }

        bool isValid(int* counts) {
            for (int i = 0; i < 10; i++) {
                if ((counts[i] != i) && (counts[i] != 0)) {
                    return false;
                }
            }
            return true;
        }

        void increment(int* digits, int* counts) {
            if (counts[*digits] != 0) {
                counts[*digits]--;
            }
            if (*digits != 9) {
                (*digits)++;
                counts[*digits]++;
            } else {
                *digits = 0;
                increment(digits + 1, counts);
                counts[*digits]++;
            }
        }
    };

File: leetcode/balanced_number.py
class Solution:
    def nextBeautifulNumber(self, n: int) -> int:
        ds = self.digitize(n)
        num_first = self.count(ds, ds[0])
        # n = aabd
        # Add 1 until there are "a" "a"s
        # then check the rest
        # then add 1 until there are "a" "a"s

        return num_first


    def digitize(self, n: int) -> List[int]:
        digits = []
        while n:
            digits.insert(0, n % 10)
            n //= 10
        return digits

    # def count(self, ds, digit):
    #     return sum(d == digit for d in ds)

    def is_balanced(self, ds):
        cs = self.count_all(ds)
        return all(k == v for k,v in cs.items())

    def count_all(self, ds):
        cs = {}
        for d in ds:
            if d in cs:
                cs[d] += 1
            else:
                cs[d] = 1
        return cs

    def increment(self, ds):
        if ds[0] != 9:
            ds[0] += 1
        else:

        for i in range(len(ds)):
            if ds
